### PR TITLE
weak memoize passed validateOptions(), drop support for schema as str

### PR DESCRIPTION
**Pull request by hulkish**
### Notable Changes

This change aims to reduce extraneous runs of `validateOptions()`. I believe this will be especially beneficial to loaders. For ex: https://github.com/webpack-contrib/style-loader/blob/master/index.js#L17.

**Update:** This pr includes a breaking change. However, since `schema-utils` is primarily used for core webpack libs - this should allow consumers to avoid major bump, since its impl is entirely internal.

#### `Commit Message Summary (CHANGELOG)`

```bash
PERF: weak memoize passed `validateOptions()`
BREAKING CHANGE: drop support for typeof schema === 'string'
```

### Type

- [X] Perf

### Issues

N/A

### SemVer

- [X] Breaking Change (:label: Major)

### Checklist

- [X] I have signed the CLA
- [X] Lint and unit tests pass with my changes
- [X] I have added tests that prove my fix is effective/works (if needed)
- [X] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes are merged and published in downstream modules - N/A

### CODE REVIEW

File: test/index.test.js [hunk 1]
```diff
@@ -48,6 +73,12 @@ describe('Error', () => {
     return validateOptions('test/fixtures/schema.json', options, '{Name}');
   };
 
+  test('should throw non objects', () => {
```
Comment (hulkish) [inline test/index.test.js]: @evilebottnawi @sokra I'd like to verify that this is not a breaking change.

Comment (evilebottnawi) [reply test/index.test.js]: @hulkish before it was throw other value?

Comment (hulkish) [reply test/index.test.js]: same test case for prior code yields:
```
 FAIL  test/index.test.js
  Valid
    ✓ should pass, uncached (7ms)
    ✕ should pass, cached
  Error
    ✓ should throw non objects (1ms)
    ✓ Throws (1ms)
    ✓ Errors (3ms)
    Messages
      ✓ Default
      ✓ Customized (4ms)

  ● Valid › should pass, cached

    SyntaxError: Unexpected token u in JSON at position 0
        at JSON.parse (<anonymous>)

      at validateOptions (src/validateOptions.js:28:19)
      at validate (test/index.test.js:31:12)
      at Object.test (test/index.test.js:51:12)
          at new Promise (<anonymous>)
          at <anonymous>
```

**Comment by hulkish** [comment]
Have also verified tests pass on `node@4.0.0`:
```
~/dev/repos/schema-utils$ nvm use 4
Now using node v4.0.0 (npm v2.14.2)
~/dev/repos/schema-utils$ node -v
v4.0.0
~/dev/repos/schema-utils$ ./node_modules/.bin/jest
 PASS  test/index.test.js
  Valid
    ✓ should pass, uncached (9ms)
    ✓ should pass, cached (1ms)
  Error
    ✓ should throw non objects (1ms)
    ✓ Throws (3ms)
    ✓ Errors (3ms)
    Messages
      ✓ Default (1ms)
      ✓ Customized (6ms)

Test Suites: 1 passed, 1 total
Tests:       7 passed, 7 total
Snapshots:   6 passed, 6 total
Time:        0.918s, estimated 1s
Ran all test suites.
```

**Comment by hulkish** [comment]
Also, a benchmark:

```js
const Benchmark = require('benchmark');

const validateOptions = require('./src/validateOptions');

const validOptions = {
  type() {},
  array: ['a'],
  string: 'hello',
  object: {
    prop: false,
    object: {
      prop: false,
    },
  },
  boolean: true,
  instance: new RegExp(''),
};

Benchmark.options.minSamples = 100;

new Benchmark.Suite()
  .add({
    name: 'validateOptions.old()',
    fn() {
      validateOptions.old('test/fixtures/schema.json', validOptions, '{Name}');
    },
  })
  .add({
    name: 'validateOptions()',
    fn() {
      validateOptions('test/fixtures/schema.json', validOptions, '{Name}');
    },
  })
  .on('cycle', (event) => {
    console.log(String(event.target));
  })
  .on('complete', function() {
    console.log('Fastest is', this.filter('fastest').map('name'));
  })
  .run();
```

```sh
~/dev/repos/schema-utils$ node benchmark.js
validateOptions.old() x 12,390 ops/sec ±1.11% (184 runs sampled)
validateOptions() x 81,615,545 ops/sec ±0.66% (187 runs sampled)
Fastest is [ 'validateOptions()' ]
```

**Comment by evilebottnawi** [comment]
@hulkish AWESOME!

File: src/validateOptions.js [hunk 2]
```diff
@@ -22,17 +22,24 @@ const ajv = new Ajv({
...
 keywords(ajv, ['instanceof', 'typeof']);
 
+const passed = new WeakSet();
+
 const validateOptions = (schema, options, name) => {
-  if (typeof schema === 'string') {
-    schema = fs.readFileSync(path.resolve(schema), 'utf8');
-    schema = JSON.parse(schema);
-  }
+  if (!passed.has(options)) {
```
Comment (sokra) [inline src/validateOptions.js]: `schema` need to be stored too.

`WeakMap<Schema, WeakSet<Options>>`

Note that Weak doesn't allow string as key. Best skip cache for string values. We can't be sure that the same string results in the same Schema. One could change the file between calls.

I would recommend to remove file reading in the next major version.

Comment (hulkish) [reply src/validateOptions.js]: agreed, since its usage is loader author facing

Comment (hulkish) [reply src/validateOptions.js]: @sokra hmm, i just pushed a similar solution for this - i think this covers when schema is string.

Comment (hulkish) [reply src/validateOptions.js]: @sokra I've gone ahead and applied these changes - which has upgraded this pr to a breaking change. My thinking is this could go in next major release.

Comment (hulkish) [reply src/validateOptions.js]: hmm... shouldn't just marking the options object passed be enough?

File: src/validateOptions.js [hunk 3]
```diff
@@ -22,17 +19,29 @@ const ajv = new Ajv({
...
+    }
+
+    passedBySchema.set(schema, new WeakSet([options]));
   }
 
   return true;
 };
 
 module.exports = validateOptions;
+module.exports.ajv = ajv;
```
Comment (sokra) [inline src/validateOptions.js]: Why exposing ajv?

File: src/validateOptions.js [hunk 4]
```diff
@@ -22,17 +19,29 @@ const ajv = new Ajv({
 errors(ajv);
 keywords(ajv, ['instanceof', 'typeof']);
 
-const validateOptions = (schema, options, name) => {
-  if (typeof schema === 'string') {
```
Comment (sokra) [inline src/validateOptions.js]: I would recommend to move this breaking change into a separate PR. It make sense but it's not worth adding a new major version for only this change. All users would need to upgrade and not upgraded user won't benefit from the performance plus.

File: src/validateOptions.js (hunk 3 above)
Comment (hulkish) [reply src/validateOptions.js]: @sokra so it can be spied: https://github.com/webpack-contrib/schema-utils/blob/583f633d558d2e57cf272b175a3897c31fc1014d/test/index.test.js#L29

File: src/validateOptions.js (hunk 4 above)
Comment (hulkish) [reply src/validateOptions.js]: makes sense

File: src/validateOptions.js (hunk 3 above)
Comment (evilebottnawi) [reply src/validateOptions.js]: @hulkish it is possible spy without export?

**Comment by sokra** [comment]
Maybe you don't like to hear this, but I think we should flag this as breaking change. The caching can lead to incorrect results in cases when an object is mutated:

``` js
validateOptions(schema, options); // ok
options.something = "invalid";
validateOptions(schema, options); // ok because cached, but should fail instead.
```

**Comment by hulkish** [comment]
after discussing with @mikesherov, i think the benefit this change offers does not justify risking correctness
